File: app/src/main/java/me/imzack/app/ender/presenter/PlanSearchPresenter.java
package me.imzack.app.ender.presenter;

import android.text.TextUtils;

import me.imzack.app.ender.model.DataManager;
import me.imzack.app.ender.model.bean.Plan;
import me.imzack.app.ender.view.adapter.PlanSearchListAdapter;
import me.imzack.app.ender.view.contract.PlanSearchViewContract;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class PlanSearchPresenter extends BasePresenter {

    private PlanSearchViewContract mPlanSearchViewContract;
    private DataManager mDataManager;
    private List<Plan> mPlanSearchList;
    private PlanSearchListAdapter mPlanSearchListAdapter;

    @Inject
    PlanSearchPresenter(PlanSearchViewContract planSearchViewContract, DataManager dataManager) {
        mPlanSearchViewContract = planSearchViewContract;
        mDataManager = dataManager;

        mPlanSearchList = new ArrayList<>();

        mPlanSearchListAdapter = new PlanSearchListAdapter(mDataManager, mPlanSearchList);
        mPlanSearchListAdapter.setOnPlanItemClickListener(new PlanSearchListAdapter.OnPlanItemClickListener() {
            @Override
            public void onPlanItemClick(int planListPos) {
                mPlanSearchViewContract.onPlanItemClicked(planListPos);
            }
        });
    }

    @Override
    public void attach() {
        mPlanSearchViewContract.showInitialView(mDataManager.getPlanCount(), mPlanSearchListAdapter);
    }

    @Override
    public void detach() {
        mPlanSearchViewContract = null;
    }

    public void notifySearchTextChanged(String searchText) {
        mDataManager.searchPlan(mPlanSearchList, searchText);
        mPlanSearchListAdapter.notifyDataSetChanged();
        mPlanSearchViewContract.onSearchChanged(TextUtils.isEmpty(searchText), mPlanSearchList.size() == 0);
    }
}

File: app/src/main/kotlin/me/imzack/app/end/presenter/PlanSearchPresenter.kt
package me.imzack.app.end.presenter

import android.text.TextUtils

import me.imzack.app.end.model.DataManager
import me.imzack.app.end.model.bean.Plan
import me.imzack.app.end.view.adapter.PlanSearchListAdapter
import me.imzack.app.end.view.contract.PlanSearchViewContract

import javax.inject.Inject

class PlanSearchPresenter @Inject constructor(
        private var mPlanSearchViewContract: PlanSearchViewContract?
) : BasePresenter() {

    private val mPlanSearchList = mutableListOf<Plan>()
    private val mPlanSearchListAdapter = PlanSearchListAdapter(mPlanSearchList)

    override fun attach() {
        mPlanSearchListAdapter.mOnPlanItemClickListener = { mPlanSearchViewContract!!.onPlanItemClicked(it) }

        mPlanSearchViewContract!!.showInitialView(DataManager.planCount, mPlanSearchListAdapter)
    }

    override fun detach() {
        mPlanSearchViewContract = null
    }

    fun notifySearchTextChanged(searchText: String) {
        DataManager.searchPlan(mPlanSearchList, searchText)
        mPlanSearchListAdapter.notifyDataSetChanged()
        mPlanSearchViewContract!!.onSearchChanged(TextUtils.isEmpty(searchText), mPlanSearchList.size == 0)
    }
}
